### ISSUE TITLE
Sopel gets stuck when attempting to connect

### ISSUE THREAD

**Issue by ralienpp** [sopel-irc/sopel#1393]
I ran into an issue related to connection behaviour, here is a log excerpt that reflects the situation:

```
pi@localhost:~$ journalctl -u sopel -f
-- Logs begin at Fri 2018-09-14 00:35:15 UTC. --
Oct 02 13:06:48 localhost python[455]: Welcome to Sopel. Loading modules...
Oct 02 13:06:48 localhost python[455]: Registered 8 modules,
Oct 02 13:06:48 localhost python[455]: 0 modules failed to load
Oct 02 13:06:48 localhost python[455]: Connecting to server.com:788...
Oct 02 13:09:01 localhost python[455]: Connection error: [Errno 110] Connection timed out
Oct 02 13:09:01 localhost python[455]: Warning: Disconnected. Reconnecting in 20 seconds...
Oct 02 13:09:21 localhost python[455]: Welcome to Sopel. Loading modules...
Oct 02 13:09:22 localhost python[455]: Registered 8 modules,
Oct 02 13:09:22 localhost python[455]: 0 modules failed to load
Oct 02 13:09:22 localhost python[455]: Connecting to server.com:788...

<end of the log>
<a couple of days later I get to the machine physically to examine its state>

Oct 04 10:16:46 localhost systemd[1]: Stopping Sopel IRC bot...
Oct 04 10:16:46 localhost python[455]: Got quit signal, shutting down.
Oct 04 10:18:16 localhost systemd[1]: sopel.service stop-sigterm timed out. Killing.
Oct 04 10:18:16 localhost systemd[1]: sopel.service: main process exited, code=killed, status=9/KILL
Oct 04 10:18:16 localhost systemd[1]: Unit sopel.service entered failed state.
```

As you can see, there was a timeout at 13:09:01, and 20 seconds later Sopel attempted another connection - but nothing else happened after that. The process appears to get stuck after this line, perhaps waiting forever, as if no timeout was set?

The log is obtained from a service managed by `systemd`, which starts Sopel.  After restarting the service, Sopel was able to connect successfully.

- This is happening on `4.9.35+ #1014 Fri Jun 30 14:34:49 BST 2017 armv6l GNU/Linux`
- `sopel==6.5.3`

Given that `Got quit signal, shutting down` is shown, it means that the signal handler was triggered, so the process was not frozen https://github.com/sopel-irc/sopel/blob/a0c567c09083414d20ad4f989012f80062d6d2fa/sopel/__init__.py#L72.

However, I don't know where to go beyond that, as I am not familiar with the low-level details of what happens under the hood.

**Comment by dgw** [comment]
Has this happened more than once?

The "Warning: Disconnected" line comes from:

https://github.com/sopel-irc/sopel/blob/a0c567c09083414d20ad4f989012f80062d6d2fa/sopel/__init__.py#L103

which is inside a `while True:` loop. I have a vague theory about why this might have happened, but it's not really feasible to test unless you can reproduce the conditions that caused this hang on-demand to test different ways of handling them.

**Comment by ralienpp** [comment]
Yes, this happened more than once and on more than one system where I use Sopel.

Reliable steps of reproducing it are not known yet. What I can do, is run another version with more logging in it, such that next time it happens, we have more data to look at. 


Another thing to consider is to let it crash. At least in my case, I have systemd to monitor the service and restart it, if necessary. So a crash would be my preferred way of "dealing" with it. 

If we let it crash, the whole traceback will be printed and logged - that will tell us exactly what exception occurred.

**Comment by dgw** [comment]
I doubt it's as simple as letting Sopel crash.

If a connection attempt does throw an exception, it will either be handled—and Sopel will fall through to initiating a reconnect—or not—in which case Sopel's top-level exception handler [in `__init__.run()`](https://github.com/sopel-irc/sopel/blob/a0c567c09083414d20ad4f989012f80062d6d2fa/sopel/__init__.py#L85-L97) would print the stack trace.

If a connection attempt fails, but no exception is thrown, Sopel gets to the end of the `while True:` loop without setting any of the flags that would break it (quit-on-request, etc.) and starts over. In theory.

Since you have a number of systems that could all experience this, I made a custom branch off of the version you're running (tag `v6.5.3`) with extra debug logging as you suggested: dgw/sopel@248f6d7. You should be able to install it however you like (Git checkout, pip install, etc.) on your systems. The only changes are adding some more `stderr()` messages in the function where I think it's getting stuck.

**Comment by dgw** [comment]
@ralienpp Anything more on this?

I don't want to leave unconfirmed issues open indefinitely. So, this will be closed if we can't get any further details by the time **7.0.0** releases. (You can always reopen it, don't worry. No thread-locking! :smile_cat:)

**Comment by dgw** [comment]
@ralienpp We've refactored a ton of stuff for 7.0.0, and still more code rewrites to come. Add to that the months of silence on gathering further information, and I think it's time to close this one. As said before, you can always feel free to comment still if you have new information to share! And we can reopen this if it still happens in the released version of 7.0 (tentatively, coming early-mid December).
